mountpoint-s3/tests/fuse_tests/write_test.rs

```diff
@@ -682,3 +666,129 @@ fn multi_thread_test() {
     .join()
     .unwrap();
 }
+
+fn overwrite_test<F>(creator_fn: F, prefix: &str)
+where
+    F: FnOnce(&str, TestSessionConfig) -> (TempDir, BackgroundSession, TestClientBox),
+{
+    let filesystem_config = S3FilesystemConfig {
+        allow_overwrite: true,
+        ..Default::default()
+    };
+    let test_config = TestSessionConfig {
+        filesystem_config,
+        ..Default::default()
+    };
+    let (mount_point, _session, mut test_client) = creator_fn(prefix, test_config);
+
+    // Make sure there's an existing directory and a file
+    test_client.put_object("dir/hello.txt", b"hello world").unwrap();
+
+    let _subdir = mount_point.path().join("dir");
+    let path = mount_point.path().join("dir/hello.txt");
+
+    // Open the file with O_RDWR but do nothing
+    let mut options = File::options();
+    let write_fh = options
+        .read(true)
+        .write(true)
+        .open(&path)
+        .expect("open with O_RDWR should succeed");
+    drop(write_fh);
+
+    // Open the file with O_WRONLY but do nothing
+    let mut options = File::options();
+    let write_fh = options
+        .write(true)
+        .open(&path)
+        .expect("open with O_WRONLY should succeed");
+    drop(write_fh);
+
+    // File content should not be changed
+    let mut options = File::options();
+    let mut fh = options.read(true).write(true).open(&path).unwrap();
+    let mut hello_contents = String::new();
+    fh.read_to_string(&mut hello_contents).unwrap();
+    assert_eq!(hello_contents, "hello world");
+
+    // We can't write to the file that is being read
+    // from both the same file handle or a new one
+    let err = fh
+        .write(b"hello world")
+        .expect_err("writing to a file is being read should fail");
+    assert_eq!(err.raw_os_error(), Some(libc::EPERM));
+
+    let mut options = File::options();
+    let mut write_fh = options.write(true).open(&path).unwrap();
+    let err = write_fh
+        .write(b"hello world")
+        .expect_err("writing to a file is being read should fail");
+    assert_eq!(err.raw_os_error(), Some(libc::EPERM));
+
+    drop(fh);
+    drop(write_fh);
+
+    // Add some delay because release is asynchronous
+    std::thread::sleep(Duration::from_secs(1));
```

Comment:
We mostly removed this problem when we implemented flush. Why are we relying on release?

Reply:
I still think it makes sense that users must close the file handle before they can start the overwrite. Maybe what we should do is `finish_reading` at flush? I updated the semantics document for now.